File: benchmark/synduce/constraints/memo/tree_size.f
Inductive Tree = leaf Int | node {Int, Tree, Tree};
Inductive TreeMemo = mleaf Int | mnode {Int, Int, TreeMemo, TreeMemo};

memo = \t: TreeMemo.
  match t with
    mleaf _ -> 1
  | mnode {x, _, _, _} -> x
  end;

is_memo = fix (
  \f: TreeMemo -> Bool. \t: TreeMemo.
  match t with
    mleaf x -> true
  | mnode {n, a, l, r} -> and (== n (+ 1 (+ (memo l) (memo r)))) (and (f l) (f r))
  end
);

repr = fix (
  \f: TreeMemo -> Tree. \t: TreeMemo.
  match t with
    mleaf a -> leaf a
  | mnode {n, a, l, r} -> node {a, f l, f r}
  end
);

target = fix (
  \f: TreeMemo -> Compress TreeMemo. \t: TreeMemo.
  match t with
    mleaf a -> mleaf a
  | mnode {n, a, l, r} -> mnode {n, a, l, r} /*Avoid recursions of l and r*/
  end
);

spec = fix (
  \f: Tree -> Int. \t: Tree.
  match t with
    leaf a -> 1
  | node {a, l, r} -> + 1 (+ (f l) (f r))
  end
);

/* Customized generator, not used */
gen = fix (
  \f: Tree -> TreeMemo. \t: Tree.
  match t with
    leaf a -> mleaf a
  | node {a, l, r} ->
    let res = {f l, f r} in
      mnode {+ 1 (+ (memo res.1) (memo res.2)), a, res.1, res.2}
  end
);

main =
  \mt: TreeMemo.
    if is_memo mt then spec (repr (target mt)) else 0;

File: ml-benchmark/synduce/constraints/memo/tree_size.ml
type tree = Leaf of int | Node of int * tree * tree

type treememo = Mleaf of int | Mnode of int * int * treememo * treememo

let memo t =
  match t with
  | Mleaf _ -> 1
  | Mnode (x, _, _, _) -> x

let rec is_memo t =
  match t with
  | Mleaf x -> true
  | Mnode (n, a, l, r) ->
      (n = 1 + memo l + memo r) && (is_memo l && is_memo r)

let rec repr t =
  match t with
  | Mleaf a -> Leaf a
  | Mnode (n, a, l, r) -> Node (a, repr l, repr r)

let target t =
  match t with
  | Mleaf a -> Mleaf a
  | Mnode (n, a, l, r) -> Mnode (n, a, l, r)

let rec spec t =
  match t with
  | Leaf a -> 1
  | Node (a, l, r) -> 1 + spec l + spec r

let rec gen t =
  match t with
  | Leaf a -> Mleaf a
  | Node (a, l, r) ->
      let res = (gen l, gen r) in
      match res with
      | (l1, r1) -> Mnode (1 + memo l1 + memo r1, a, l1, r1)

let program mt =
  if is_memo mt then spec (repr (target mt)) else 0